azdev/operations/code_gen.py

```diff
@@ -300,3 +306,81 @@ def _create_package(prefix, repo_path, is_ext, name='test', display_name=None, d
+            (swagger_repo.startswith('https://') and swagger_repo.endswith(const.SWAGGER_REPO_NAME)):
+        swagger_readme_file_path = '{}/blob/{}/specification/{}/resource-manager'.format(
+            swagger_repo, branch, ext_name)
+        # validate URL
+        try:
+            request.urlopen(swagger_readme_file_path)
+        except error.HTTPError as ex:
+            raise CLIError('HTTPError: {}\nThe URL {} does not exist.'.format(ex.code, swagger_readme_file_path))
+    else:
+        swagger_readme_file_path = os.path.join(swagger_repo, 'specification', ext_name, 'resource-manager')
+        if not os.path.isdir(swagger_readme_file_path):
+            raise CLIError("The path {} does not exist.".format(swagger_readme_file_path))
+    return swagger_readme_file_path
+
+
+def _generate_extension(ext_name, repo_path, swagger_readme_file_path, use):
+    heading('Start generating extension {}.'.format(ext_name))
+    # check if npm is installed
+    try:
+        subprocess.run('npm --version', shell=True, check=True, stdout=subprocess.DEVNULL)
+    except subprocess.CalledProcessError as ex:
+        raise CLIError('{}\nPlease install npm.'.format(ex))
+    display('Installing autorest.\n')
+    if const.IS_WINDOWS:
+        try:
+            subprocess.run('npm install -g autorest', shell=True, check=True)
+        except subprocess.CalledProcessError as ex:
+            raise CLIError("Failed to install autorest.\n{}".format(ex))
+    else:
+        try:
+            subprocess.run('npm install -g autorest', shell=True, check=True, stderr=subprocess.DEVNULL)
+        except subprocess.CalledProcessError as ex:
+            path = os.environ['PATH']
+            # check if npm is installed through nvm
+            if os.environ.get('NVM_DIR'):
+                raise CLIError(ex)
+            else:
+                # check if user using specific node version and manually add it to the os env PATH
+                node_version = subprocess.check_output('node --version', shell=True).decode('utf-8')
+                if node_version.endswith('\n'):
+                    node_version = node_version.replace('\n', '')
+                if 'node/' + node_version + '/bin' in path:
+                    raise CLIError(ex)
+                else:
+                    # create a new directory for npm global installations, to avoid using sudo in installing autorest
+                    npm_path = os.path.join(os.environ['HOME'], '.npm-packages')
+                    if not os.path.isdir(npm_path):
+                        os.mkdir(npm_path)
+                    npm_prefix = subprocess.check_output('npm prefix -g', shell=True).decode('utf-8')
+                    if npm_prefix.endswith('\n'):
+                        npm_prefix = npm_prefix.replace('\n', '')
+                    subprocess.run('npm config set prefix ' + npm_path, shell=True)
+                    os.environ['PATH'] = path + ':' + os.path.join(npm_path, 'bin')
```

Comment:
this is little aggressive, to change user's PATH directly. Is there any other way? at least a message to let user know PATH is changed.

Reply:
This does not actually change the path on the user's profile. It changes it only in Python's view of the OS. So, when a new process is spawned by the Linux kernel in this case python would pass the os.environ dictionary to the linux spawned new process. That process would have this modified path only, basically this spawned processes environment is what Python's os.environ dictionary is, and once it dies it will go away not effecting the user. The only thing here that could effect the outside OS is the prefix change, which is set back to its default upon completion of the autorest install.